src/ui/components/TestSuite/suspense/TestStepSourceLocationCache.ts

```diff
@@ -54,10 +54,34 @@ export const TestStepSourceLocationCache = createCacheWithTelemetry<
                 // isn't from the same source as the marker to identify the user frame
                 // that invoke it
                 return f.functionLocation?.every(fl => fl.sourceId !== markerSourceId);
-              } else {
-                // for enqueued assertions, search from the top for the first frame from the same source
-                // as the marker (which should be cypress_runner.js) and return it
+              } else if (!f.functionLocation?.some(fl => fl.sourceId === markerSourceId)) {
+                // for enqueued assertions, the source will generally be the
+                // first location that isn't the same as the marker (which is
+                // the cypress_runner).
+                //
+                // The frames array is usually one of the two:
+                // * [
+                //     addAnnotation and related plugin code,
+                //     cypress code for the cy.* command in cypress_runner,
+                //     user code,
+                //     __stackReplacementMarker in cypress_runner
+                //   ]
+                // * [
+                //     addAnnotation and related plugin code,
+                //     cypress code for the cy.* command in cypress_runner,
+                //     user code,
+                //     wrapper code like for cy.then in cypress_runner
+                //     __stackReplacementMarker in cypress_runner
+                //   ]
+                //
+                // We find it by looking from the top (the frames array comes in
+                // with the most recent last but it is reversed above) for the
+                // first entry that isn't from the cypress_runner where the
+                // _next_ frame is from the cypress_runner.
+
                 return l[i + 1]?.functionLocation?.some(fl => fl.sourceId === markerSourceId);
```

Comment:
I know you didn't newly add these variables in this PR, but I'd prefer we avoid variable names like `l` and `fl`. They're really confusing when you get nested deeply enough to not be able to see where they were defined.

Reply:
Sure. Can do!